iroh/src/endpoint/connection.rs

```diff
@@ -0,0 +1,1721 @@
+    pub fn ignore(self) {
+        self.inner.ignore()
+    }
+
+    /// Returns the local IP address which was used when the peer established the
+    /// connection.
+    pub fn local_ip(&self) -> Option<IpAddr> {
+        self.inner.local_ip()
+    }
+
+    /// Returns the peer's UDP address.
+    pub fn remote_address(&self) -> SocketAddr {
+        self.inner.remote_address()
+    }
+
+    /// Whether the socket address that is initiating this connection has been validated.
+    ///
+    /// This means that the sender of the initial packet has proved that they can receive
+    /// traffic sent to `self.remote_address()`.
+    pub fn remote_address_validated(&self) -> bool {
+        self.inner.remote_address_validated()
+    }
+}
+
+impl IntoFuture for Incoming {
+    type Output = Result<Connection, ConnectingError>;
+    type IntoFuture = IncomingFuture;
+
+    fn into_future(self) -> Self::IntoFuture {
+        IncomingFuture {
+            inner: self.inner.into_future(),
+            ep: self.ep,
+        }
+    }
+}
+
+/// Adaptor to let [`Incoming`] be `await`ed like a [`Connecting`].
+#[derive(Debug)]
+#[pin_project]
+pub struct IncomingFuture {
+    #[pin]
+    inner: quinn::IncomingFuture,
+    ep: Endpoint,
+}
+
+impl Future for IncomingFuture {
+    type Output = Result<Connection, ConnectingError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        match this.inner.poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Err(err)) => Poll::Ready(Err(err.into())),
+            Poll::Ready(Ok(inner)) => {
+                let conn = match conn_from_quinn_conn(inner) {
+                    Ok(conn) => conn,
+                    Err(err) => return Poll::Ready(Err(HandshakeFailureSnafu.into_error(err))),
+                };
+                try_send_rtt_msg(conn.quinn_connection(), this.ep, Some(conn.remote_id()));
+                Poll::Ready(Ok(conn))
+            }
+        }
+    }
+}
+/// Extracts the ALPN protocol from the peer's handshake data.
+fn alpn_from_quinn_conn(conn: &quinn::Connection) -> Option<Vec<u8>> {
+    let data = conn.handshake_data()?;
+    match data.downcast::<quinn::crypto::rustls::HandshakeData>() {
+        Ok(data) => data.protocol,
+        Err(_) => None,
+    }
+}
+
+async fn alpn_from_quinn_connecting(conn: &mut quinn::Connecting) -> Result<Vec<u8>, AlpnError> {
+    let data = conn.handshake_data().await?;
+    match data.downcast::<quinn::crypto::rustls::HandshakeData>() {
+        Ok(data) => match data.protocol {
+            Some(protocol) => Ok(protocol),
+            None => Err(UnavailableSnafu.build()),
+        },
+        Err(_) => Err(UnknownHandshakeSnafu.build()),
+    }
+}
+
+#[allow(missing_docs, dead_code)]
+#[common_fields({
+    backtrace: Option<snafu::Backtrace>,
+})]
+#[derive(Debug, Snafu)]
+#[non_exhaustive]
+pub enum QuinnConnectionError {
```

Comment:
Would love if we could rename this. It has to be exposed to the user, but the user might not really know what "Quinn" is.

Reply:
I know we like to make all errors as specific as possible, but at this point we have `RemoteEndpointIdError`, `QuinnConnectionError` and `ConnectingError`.
Would it be insane to consolidate all of them into one `HandshakingError` or something similar? What do others think?

Reply:
Or maybe just `ConnectError`?